### PR TITLE
Added OpenKilda contribution workflow description

### PR DESCRIPTION
**Pull request by niksv**
This commit introduces a contribution guide to assist developers in contributing to the Open-Kilda project in accordance with its guidelines.

### CODE REVIEW

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:51]: I'd prefer that branch name start from issue ID and to be as close to issue description as possible (of course, trying to make it short). E.g. `feature/1535-add-wednesday-reminder`

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:154]: Let's use `test/` directory, as mentioned in `Branch name` part of this guide.

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:178]: What about test PRs? Should we follow the same labeling policy (adding tests = `feature`, fixing tests = `bugfix`, adding new helper methods/refactoring = `improvement`)? If so, let's add this to the guide.

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:180]: Should test PR inherit area from the tested feature and add it's own `area/testing` label or only use `area/testing` one?

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:183]:
```
...you may use label `Draft` to mark such a PR...
```

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:189]: s/RP/PR/

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:190]: You can't finish the phrase like this. Add the fatality:
```
... Deal with it 😎
```

File: docs/contrib/workflow.md
Comment (pkazlenka) [inline docs/contrib/workflow.md:207]: We miss the step where autotests are merged to the feature branch (unless you prefer to merge them later to `develop`).
The easier to to follow algorithm is to merge tests into feature branch after it got `Ready to Merge` label.
As an option, we can say that tests can be merged at any moment when tester believes that the tests cover the feature and use the right APIs. But in this case the later maintenance and functional test runs against the feature/bugfix would become a developer's task.

**Review by pkazlenka — CHANGES_REQUESTED** [review]
Please, fix several misprints and let's discuss in comments some testing-related rules.

File: docs/contrib/workflow.md
Comment (dmitrii-beliakov) [inline docs/contrib/workflow.md:222]: Is it really needed? Seems not applicable to a regular contributor. Below you write that only those with the rights can merge to develop. It's not `do not merge`, it's in fact `you don't have permissions to merge`.

File: docs/contrib/workflow.md
Comment (dmitrii-beliakov) [inline docs/contrib/workflow.md:221]: Is there any rules for deleting branches after merging? Any rules for squashing commits? Is it actually rebasing or merging?
At least the deletion policy would be useful. Does setting `ready to merge` also imply `say good bye to your branch`? Can disappear at any moment or will be there forever?

File: docs/contrib/workflow.md
Comment (dmitrii-beliakov) [inline docs/contrib/workflow.md:119]: Is it? This PR doesn't have a description, but the first commit has a message `Added OpenKilda contribution workflow description`

File: docs/contrib/workflow.md
Comment (dmitrii-beliakov) [inline docs/contrib/workflow.md:172]: _a guy who submitted 683 lines long unit test class sad noises_ 
Should it be some artificial separation or let's say based on different packages? For example, I can imagine I split my PR into PR with docs, PR with API contract, PR with impl. But should I split impl and unit tests for that impl?

File: docs/contrib/workflow.md
Comment (pablomuri) [inline docs/contrib/workflow.md:115]: Typo: several sentences which **describe** your changes and **give** some additional information. 
I would add something like "Try to explain what you do, and why are you doing it.". It will also help see if the developer actually understands the reason for the change. Sometimes it is just understanding the bug.

File: docs/contrib/workflow.md
Comment (pablomuri) [reply docs/contrib/workflow.md:119]: That's true. Github actually takes the description of the commit. But this commit has no description, only a title.
@niksv I would add a description as well to the commit. "This commit introduces a contribution guide to assist developers in contributing to the Open-Kilda project in accordance with its guidelines." (I am starting to love AI's btw)

File: docs/contrib/workflow.md
Comment (pablomuri) [inline docs/contrib/workflow.md:136]: Why only the first one? Because the PR has going to be squashed?

File: docs/contrib/workflow.md
Comment (pablomuri) [inline docs/contrib/workflow.md:152]: typo: A developer has **a** branch `feature/ha_flows` with a new feature and wants to create a PR.

File: docs/contrib/workflow.md
Comment (pablomuri) [inline docs/contrib/workflow.md:153]: I'd say "The developer" instead of "he/she"

File: docs/contrib/workflow.md
Comment (pablomuri) [inline docs/contrib/workflow.md:155]: I'd say "The developer" instead of "he/she"

The developer creates **a** new PR from **the** branch `tests/ha_flows_func_tests` with the base branch

File: docs/contrib/workflow.md
Comment (pablomuri) [reply docs/contrib/workflow.md:172]: totally agree

**Review by pablomuri — CHANGES_REQUESTED** [review]
Great job @niksv. This will be very useful for the team.

Regarding the labels, we could also add them to the commit title something like (feat, fix, test, docs...). For example with the feat you can do something like: 
feat: new thing
feat(ha_flows): add API objects
feat(ha_flows): implement create operation
feat(ha_flows): implement delete operation

See https://www.conventionalcommits.org/en/v1.0.0/

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md:207]: There are several problems with merging tests into feature branch
1. We are loosing test discussion history. Merged code will point on feature PR, but all tests discussion will be on test PR
2. We can loose test PR in release notes. Sometimes PRs breaks something in release. In this case to check what was included in release a release manages checks release notes. If a release manager will merge only PR witth feature and tests - the manager will loose a link to tests PR.

The general approach is:
1. release manager merge feature into develop
2. deletes feature branch
3. GitHub changes base branch of PR with tests to `develop` automatically.
4. Release manager merge tests into develop

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md:221]: Usually I delete a branch after merge. Do we have some reasons not to do it?

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md:222]: You do not have rights to merge in develop, but you have rights to merge in other branches. For example if your PR is based on some other PR - you can merge your code into the branch with some other PR

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md:231]: internal contributor can just ask the team. I'm not sure that external contributors can merge a PR

**Review by pkazlenka — APPROVED** [review]

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md:125]: added)

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md:158]: why **a**? branch has a name. @dmitrii-beliakov we need your help here

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md]: fixed

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md]: fixed

File: docs/contrib/workflow.md
Comment (pablomuri) [reply docs/contrib/workflow.md:158]: Actually I am not sure. Maybe you can say "A developer has a branch called `feature/ha_flows` with a new feature and wants to create a PR."

File: docs/contrib/workflow.md
Comment (dmitrii-beliakov) [reply docs/contrib/workflow.md:231]: I don't know. My comment is mostly about expectations: after somebody puts a label `ready to merge` they are not supposed to make any new commits or use that branch in some way. If there is this agreement/expectation, it's worth to write that down.

File: docs/contrib/workflow.md
Comment (dmitrii-beliakov) [reply docs/contrib/workflow.md:77]: To summarize. Having one-commit branches or changing the commits from the past brings several problems in using git and doing code review. Having more commits allows cherry-pick, revert, or just understand better, or save some time (like no need for multiple people really review commit "fix check style"). On the other hand, rebasing with multiple commits leads to more conflicts resolution. Multiple commits introduce a noise and not informative in git blame or history. 
So a compromise I see could be: squash commits before putting `ready to merge`. And the guide would be: in your feature branch do whatever you believe and agreed with collaborators is best, one or more commits, rebasing, merging, whatever. But when the work is done, a dev, responsible for a particular branch, squashes all commits into a single one, puts a summarizing message into this commit, and puts a label `ready to merge`. This way, the release engineer have a single-commit branch to merge.
I tried to do this with one of the branches: `fix/4984-sub-flow-descriptions`, you can check whether it is a good compromise. 
Here are the steps how to do that inside a feature branch:

_update all branches, using your favorite way, or:_
```
git fetch
git switch develop
git pull
git switch FEATURE_BRANCH
git pull
```
_squash the commits. The reset command reads all commits and creates a new CL as not committed changes:_
```
git switch FEATURE_BRANCH
git reset --soft $(git merge-base develop HEAD)
git commit -m "A nice summary commit message that will remain in the develop branch"
```
_submit changes to remote. It breaks history, so other people are not supposed to work on this branch at this moment:_
```
git push --force
```
_Optionally, verify how many commits are there not counting the merges:_
```
$ git rev-list --no-merges --count HEAD ^develop
1
```
_In this example there is just one commit in the current branch._

File: docs/contrib/workflow.md
Comment (dmitrii-beliakov) [inline docs/contrib/workflow.md:24]: I think there is a few points missing here: 
- we need the design doc to be reviewed, challenged, understood in terms of scope, testing, relation to other features, and so on. Before starting with the implementation, it is required get approvals. A person could cross fingers instead of creating a doc for a small change, but they might receive a comment that requests changes and it might require to do much more work comparing that if the same comment received during design doc review.
- when even a small fix makes design documentation incorrect, then either that doc must be corrected accordingly or an issue is created (maybe it is enough just to specify the doc and the issue).
- the life cycle of docs change is pretty similar to other code changes (PRs, labels ready for test and all other stuff)

File: docs/contrib/workflow.md
Comment (pkazlenka) [reply docs/contrib/workflow.md]: Gin Hub is fine too ^^

File: docs/contrib/workflow.md
Comment (niksv) [reply docs/contrib/workflow.md:231]: added

File: docs/contrib/workflow.md
Comment (pablomuri) [reply docs/contrib/workflow.md:78]: It's been almost a year. I've started to re-review the PR after the push --force and now it's difficult to understand the changes. I've tried comparing the commit id's but since we have introduced a lot of changes in the repository it's difficult to see the changes in the PR now.  I agree with @dmitrii-beliakov now. I think the best option is that just before the merge we PR owner must squash the commits into one (or two/three if the commits make sense by themselves).

File: docs/contrib/workflow.md
Comment (pablomuri) [inline docs/contrib/workflow.md:46]: I think it is better to use "docs" instead of "doc.

File: docs/contrib/workflow.md
Comment (pablomuri) [inline docs/contrib/workflow.md:52]: I would add another one:

* `chore/` - for version upgrades, bump up, formatting, cleaning useless code, etc.

Comment (niksv) [reply docs/contrib/workflow.md:52]: added

**Review by pablomuri — APPROVED** [review]

**Pull request merged**
